src/Compression/CompressionCodecEncrypted.h

```diff
@@ -1,104 +1,114 @@
+    * you want to apply both compression and encryption to your
+    * columns, you need to put this codec at the end of the chain
+    * like "column Int32 Codec(Delta, LZ4,
+    * AES_128_GCM_SIV)".
+    *
+    * The key is obtained by executing a command specified in the
+    * configuration file at startup, and if it doesn't specify a
+    * command the codec refuses to process any data. The command is
+    * expected to write a Base64-encoded key of any length, and we
+    * apply HKDF-SHA-256 to derive a 128-bit key-generation key
+    * (only the first half of the result is used). We then encrypt
+    * blocks in AES-128-GCM-SIV with a universally fixed nonce (12
+    * repeated NUL characters).
+    *
+    * This construct has a weakness due to the nonce being fixed at
+    * all times: when the same data block is encrypted twice, the
+    * resulting ciphertext will be exactly the same. We have to live
+    * with this weakness because ciphertext must be deterministic,
+    * as otherwise our engines like ReplicatedMergeTree cannot
+    * deduplicate data blocks.
+    */
+class CompressionCodecEncrypted : public ICompressionCodec
+{
+public:
+    /** If a key is available, the server is supposed to
+        * invoke this static method at the startup. The codec will
+        * refuse to compress or decompress any data until that. The
+        * key can be an arbitrary octet string, but it is
+        * recommended that the key is at least 16 octets long.
+        *
+        * Note that the key is currently not guarded by a
+        * mutex. This method should be invoked no more than once.
+        */
+    explicit CompressionCodecEncrypted(CompressionMethodByte Algorithm);
+
+    /**
+        * This is utility class. It holds information about encryption configuration.
+        */
+    class Configuration
     {
     public:
-        /** If a master key is available, the server is supposed to
-          * invoke this static method at the startup. The codec will
-          * refuse to compress or decompress any data until that. The
-          * key can be an arbitrary octet string, but it is
-          * recommended that the key is at least 16 octets long.
-          *
-          * Note that the master key is currently not guarded by a
-          * mutex. This method should be invoked no more than once.
-          */
-        static void setMasterKey(const std::string_view & master_key);
-
-        CompressionCodecEncrypted(const std::string_view & cipher);
+        static Configuration & instance();
 
-        uint8_t getMethodByte() const override;
-        void updateHash(SipHash & hash) const override;
+        void load(const Poco::Util::AbstractConfiguration & config, const String & config_prefix);
+        String getKey(UInt64 id) const;
+        String getCurrentKey() const;
+        UInt64 getCurrentKeyID() const;
+        String getNonce() const;
+        CompressionMethodByte getAlgorithmByte() const;
```

Comment:
If the code calls `getKey()` and after that `getNonce()`, the configuration could change between those calls providing inconsistent and wrong results of encryption or decryption. It would be better to provide functions which will extract everything we need to encrypt or decrypt with just one call, like

```
void getCurrentKeyAndNonce(EncryptionMethod method, UInt64 & current_key_id, String & current_key, String & nonce) const;
```